File: OJ_Assignment/Ch06-Binary_Tree/1037.6.37.cpp
#include <iostream>
#include <string>
#include <stack>
#include <vector>
using namespace std;

void PreOrderPrint(size_t index, const vector<string> &value,
                   const vector<int> &left, const vector<int> &right) {
    stack<size_t> stk;
    size_t p = index;
    while (p != -1 || !stk.empty()) {
        if (p != -1) {
            cout << value[p] << " ";
            stk.push(p);
            p = left[p];
        } else {
            p = stk.top();
            stk.pop();
            p = right[p];
        }
    }
    cout << endl;
}

int main() {
    vector<string> val(1, "");
    vector<int> left(1, 0);
    vector<int> right(1, 0);

    string str;
    int lft, rgt;
    while (cin >> str >> lft >> rgt) {
        val.push_back(str);
        left.push_back(lft);
        right.push_back(rgt);
    }

    PreOrderPrint(1, val, left, right);
    return 0;
}

File: OJ_Assignment/Ch06-Binary_Tree/1037.6.37.py
class stack:
    def __init__(self):
        self.base = []
    def push(self, data):
        self.base.append(data)
    def pop(self):
        return self.base.pop()
    def top(self):
        return self.base[len(self.base) - 1]
    def empty(self):
        return len(self.base) == 0

def PreOrderPrint(root, value = [], left = [], right = []):
    stk = stack()
    p = root
    while p != -1 or not stk.empty():
        if p != -1:
            print(value[p], end=" ")
            stk.push(p)
            p = left[p]
        else:
            p = stk.pop()
            p = right[p]
    print()

value = [0]
left = [0]
right = [0]
try:
    while True:
        line = input()
        inputList = line.split()
        value.append(inputList[0])
        left.append(int(inputList[1]))
        right.append(int(inputList[2]))
except EOFError:
    pass
PreOrderPrint(1, value, left, right)
